pkg/dataobj/internal/dataset/dataset.go

```
@@ -1,3 +1,82 @@
 // Package dataset contains utilities for working with datasets. Datasets hold
 // columnar data across multiple pages.
 package dataset
+
+import (
+	"context"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/result"
+)
+
+// A Dataset holds a collection of [Columns], each of which is split into a set
+// of [Pages] and further split into a sequence of [Values].
+//
+// Dataset is read-only; callers must not modify any of the values returned by
+// methods in Dataset.
+type Dataset interface {
+	// ListColumns returns the set of [Column]s in the Dataset. The order of
+	// Columns in the returned sequence must be consistent across calls.
+	ListColumns(ctx context.Context) result.Seq[Column]
+
+	// ListPages retrieves a set of [Pages] given a list of [Column]s.
+	// Implementations of Dataset may use ListPages to optimize for batch reads.
+	// The order of [Pages] in the returned sequence must match the order of the
+	// columns argument.
+	ListPages(ctx context.Context, columns []Column) result.Seq[Pages]
+
+	// ReadPages returns the set of [PageData] for the specified slice of pages.
+	// Implementations of Dataset may use ReadPages to optimize for batch reads.
+	// The order of [PageData] in the returned sequence must match the order of
+	// the pages argument.
+	ReadPages(ctx context.Context, pages []Page) result.Seq[PageData]
+}
+
+// FromMemory returns an in-memory [Dataset] from the given list of
+// [MemColumn]s.
+func FromMemory(columns []*MemColumn) Dataset {
+	return memDataset(columns)
+}
+
+type memDataset []*MemColumn
+
+func (d memDataset) ListColumns(_ context.Context) result.Seq[Column] {
+	return result.Iter(func(yield func(Column) bool) error {
+		for _, c := range d {
+			if !yield(c) {
+				return nil
+			}
+		}
+
+		return nil
+	})
+}
+
+func (d memDataset) ListPages(ctx context.Context, columns []Column) result.Seq[Pages] {
```

if memDataSet hold columns why do we need to pass columns ? can we just pass indexes ? Just a question really nothing important.

---

Since ListColumns must return columns in the same order every time, we definitely could just pass indexes here for any implementation of Dataset.

However, I think indexes are easier to accidentally misuse and lead to bugs that are hard to track down. One might pass around the same index set `[]int{0, 1, 2}` unintentionally to two different Datasets that return two different sets of columns. 

Having callers pass around Column instances avoids that concern; the Dataset implementation can fail early if it was given a Column that doesn't belong to it. You can see an example of me doing this in [my second prototype]( https://github.com/rfratto/loki/blob/4b759173c76c0553146cff3beee05ea880d78c22/pkg/dataobj/internal/encoding/obj/dataset_streams.go#L63-L66), though it's currently only partially handling incorrect columns.